// @lwrjs/loader/build/cjs/modules/lwr/esmLoader/importResolver.cjs
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __markAsModule = (target) => __defProp(target, "__esModule", {value: true});
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, {get: all[name], enumerable: true});
};
var __exportStar = (target, module2, desc) => {
  if (module2 && typeof module2 === "object" || typeof module2 === "function") {
    for (let key of __getOwnPropNames(module2))
      if (!__hasOwnProp.call(target, key) && key !== "default")
        __defProp(target, key, {get: () => module2[key], enumerable: !(desc = __getOwnPropDesc(module2, key)) || desc.enumerable});
  }
  return target;
};
var __toModule = (module2) => {
  return __exportStar(__markAsModule(__defProp(module2 != null ? __create(__getProtoOf(module2)) : {}, "default", module2 && module2.__esModule && "default" in module2 ? {get: () => module2.default, enumerable: true} : {value: module2, enumerable: true})), module2);
};

// packages/@lwrjs/loader/src/modules/lwr/esmLoader/importResolver.ts
__markAsModule(exports);
__export(exports, {
  default: () => importResolver_default
});
var import_types = __toModule(require("@lwrjs/types"));
var ImportResolver = class {
  constructor(config) {
    this.importURICache = new Map();
    this.modifiers = "";
    this.normalizeMetadata(config);
    this.mappingEndpoint = config?.importMappings ? void 0 : config?.endpoints?.uris.mapping;
    if (config?.endpoints?.modifiers) {
      this.modifiers = Object.entries(config.endpoints.modifiers).reduce((q, [k, v]) => q += `${k}=${v}&`, "?");
    }
  }
  normalizeMetadata(importMetada) {
    if (importMetada && importMetada.imports) {
      for (const [uri, value] of Object.entries(importMetada.imports)) {
        if (uri && value) {
          const specifiers = Array.isArray(value) ? value : [];
          specifiers.forEach((specifier) => {
            this.importURICache.set(specifier, uri);
          });
        }
      }
    }
  }
  async fetchMappings(specifier) {
    const mappingUri = `${this.mappingEndpoint}${encodeURIComponent(specifier)}${this.modifiers}`;
    const res = await globalThis.fetch(mappingUri);
    if (res.ok) {
      const mappings = await res.json();
      this.normalizeMetadata(mappings);
    }
  }
  async resolve(specifier) {
    let uri = this.importURICache.get(specifier);
    if (!uri && this.mappingEndpoint) {
      await this.fetchMappings(specifier);
      uri = this.importURICache.get(specifier);
    }
    return uri;
  }
};
var importResolver_default = ImportResolver;
