// lightning-base-components/src/lightning/confirm/confirm.js
import { api } from 'lwc';
import LightningInteractiveDialogBase from 'lightning/interactiveDialogBase';
import { parent, instanceName, secure } from 'lightning/overlayUtils';
import LightningOverlay from 'lightning/overlay';
import labelDefault from '@salesforce/label/LightningConfirm.defaultLabel';

/**
 * Create a confirm modal within your component that asks the user
 * to respond before they continue.
 */

export default class LightningConfirm extends LightningOverlay {
    static [instanceName] = 'lightning-confirm';
    static [parent] = LightningInteractiveDialogBase;

    /**
     * value to use for header text in "header" variant
     * or aria-label in "headerless" variant
     * @type {string}
     * @default "Confirm" (translated accordingly)
     */
    @api label = labelDefault;

    /**
     * text to display in body of modal
     */
    @api message = '';

    /**
     * variant to use for alert; may be
     * set to "header" or "headerless"
     */
    @api variant = 'header';

    /**
     * theme to use when variant is "header"
     * valid values are based on SLDS themes and are
     * listed in "themeOptions" of interactiveDialogBase.js
     */
    @api theme = 'default';

    /**
     * Dispatches privateclose event
     * and closes dialog
     */
    close(result) {
        const promise = new Promise((resolve) => {
            this.dispatchEvent(
                new CustomEvent('privateclose', {
                    detail: {
                        resolve,
                        [secure]: true,
                    },
                    bubbles: true,
                })
            );
        });
        super.close(result, promise);
    }

    /**
     * Dispatches privatechildrender event
     * with data parent needs to correctly render
     */
    renderedCallback() {
        const evt = new CustomEvent('privatechildrender', {
            bubbles: true,
            detail: {
                message: this.message,
                label: this.label,
                hideCancel: false,
                showDescription: true,
                getOkValue: () => true,
                getCancelValue: () => false,
                role: 'alertdialog',
            },
        });
        this.dispatchEvent(evt);
    }
}
