// @lwc/compiler/dist/index.cjs.js
/**
 * Copyright (C) 2023 salesforce.com, inc.
 */
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

var path = require('path');
var shared = require('@lwc/shared');
var errors = require('@lwc/errors');
var styleCompiler = require('@lwc/style-compiler');
var templateCompiler = require('@lwc/template-compiler');
var babel = require('@babel/core');
var babelClassPropertiesPlugin = require('@babel/plugin-proposal-class-properties');
var babelObjectRestSpreadPlugin = require('@babel/plugin-proposal-object-rest-spread');
var lwcClassTransformPlugin = require('@lwc/babel-plugin-component');
var lockerBabelPluginTransformUnforgeables = require('@locker/babel-plugin-transform-unforgeables');
var babelAsyncToGenPlugin = require('@babel/plugin-transform-async-to-generator');
var babelAsyncGeneratorFunctionsPlugin = require('@babel/plugin-proposal-async-generator-functions');

function _interopNamespaceDefault(e) {
    var n = Object.create(null);
    if (e) {
        Object.keys(e).forEach(function (k) {
            if (k !== 'default') {
                var d = Object.getOwnPropertyDescriptor(e, k);
                Object.defineProperty(n, k, d.get ? d : {
                    enumerable: true,
                    get: function () { return e[k]; }
                });
            }
        });
    }
    n.default = e;
    return Object.freeze(n);
}

var path__namespace = /*#__PURE__*/_interopNamespaceDefault(path);
var styleCompiler__namespace = /*#__PURE__*/_interopNamespaceDefault(styleCompiler);
var babel__namespace = /*#__PURE__*/_interopNamespaceDefault(babel);

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const DEFAULT_OPTIONS = {
    isExplicitImport: false,
    preserveHtmlComments: false,
    enableStaticContentOptimization: true,
    // TODO [#3370]: remove experimental template expression flag
    experimentalComplexExpressions: false,
    disableSyntheticShadowSupport: false,
    enableLightningWebSecurityTransforms: false,
};
const DEFAULT_DYNAMIC_IMPORT_CONFIG = {
    loader: '',
    strictSpecifier: true,
};
const DEFAULT_STYLESHEET_CONFIG = {
    customProperties: {
        resolution: { type: 'native' },
    },
};
const DEFAULT_OUTPUT_CONFIG = {
    minify: false,
    sourcemap: false,
};
function validateTransformOptions(options) {
    validateOptions(options);
    return normalizeOptions(options);
}
function validateOptions(options) {
    errors.invariant(!shared.isUndefined(options), errors.CompilerValidationErrors.MISSING_OPTIONS_OBJECT, [options]);
    if (!shared.isUndefined(options.enableLwcSpread)) {
        // eslint-disable-next-line no-console
        console.warn(`"enableLwcSpread" property is deprecated. The value doesn't impact the compilation and can safely be removed.`);
    }
    if (!shared.isUndefined(options.stylesheetConfig)) {
        validateStylesheetConfig(options.stylesheetConfig);
    }
    if (!shared.isUndefined(options.outputConfig)) {
        validateOutputConfig(options.outputConfig);
    }
}
function validateStylesheetConfig(config) {
    const { customProperties } = config;
    if (!shared.isUndefined(customProperties)) {
        const { resolution } = customProperties;
        if (!shared.isUndefined(resolution)) {
            errors.invariant(shared.isObject(resolution), errors.CompilerValidationErrors.INVALID_RESOLUTION_PROPERTY, [
                resolution,
            ]);
            const { type } = resolution;
            errors.invariant(type === 'native' || type === 'module', errors.CompilerValidationErrors.INVALID_TYPE_PROPERTY, [type]);
        }
    }
}
function isUndefinedOrBoolean(property) {
    return shared.isUndefined(property) || shared.isBoolean(property);
}
function validateOutputConfig(config) {
    errors.invariant(isUndefinedOrBoolean(config.sourcemap), errors.CompilerValidationErrors.INVALID_SOURCEMAP_PROPERTY, [config.sourcemap]);
    if (!shared.isUndefined(config.minify)) {
        // eslint-disable-next-line no-console
        console.warn(`"OutputConfig.minify" property is deprecated. The value doesn't impact the compilation and can safely be removed.`);
    }
}
function normalizeOptions(options) {
    const outputConfig = Object.assign(Object.assign({}, DEFAULT_OUTPUT_CONFIG), options.outputConfig);
    const stylesheetConfig = {
        customProperties: Object.assign(Object.assign({}, DEFAULT_STYLESHEET_CONFIG.customProperties), (options.stylesheetConfig && options.stylesheetConfig.customProperties)),
    };
    const experimentalDynamicComponent = Object.assign(Object.assign({}, DEFAULT_DYNAMIC_IMPORT_CONFIG), options.experimentalDynamicComponent);
    const apiVersion = shared.getAPIVersionFromNumber(options.apiVersion);
    return Object.assign(Object.assign(Object.assign({}, DEFAULT_OPTIONS), options), { stylesheetConfig,
        outputConfig,
        experimentalDynamicComponent,
        apiVersion });
}

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
function styleTransform(src, filename, config) {
    const { customProperties } = config.stylesheetConfig;
    const styleCompilerConfig = {
        customProperties: {
            resolverModule: customProperties.resolution.type === 'module'
                ? customProperties.resolution.name
                : undefined,
        },
        scoped: config.scopedStyles,
        disableSyntheticShadowSupport: config.disableSyntheticShadowSupport,
        apiVersion: config.apiVersion,
    };
    let res;
    try {
        res = styleCompiler__namespace.transform(src, filename, styleCompilerConfig);
    }
    catch (e) {
        throw errors.normalizeToCompilerError(errors.TransformerErrors.CSS_TRANSFORMER_ERROR, e, { filename });
    }
    // Rollup only cares about the mappings property on the map. Since producing a source map for
    // the styles doesn't make sense, the transform returns an empty mappings.
    return {
        code: res.code,
        map: { mappings: '' },
    };
}

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transforms a HTML template into module exporting a template function.
 * The transform also add a style import for the default stylesheet associated with
 * the template regardless if there is an actual style or not.
 */
function templateTransform(src, filename, options) {
    const { experimentalDynamicComponent, 
    // TODO [#3370]: remove experimental template expression flag
    experimentalComplexExpressions, preserveHtmlComments, enableStaticContentOptimization, customRendererConfig, enableDynamicComponents, experimentalDynamicDirective: deprecatedDynamicDirective, instrumentation, namespace, name, apiVersion, } = options;
    const experimentalDynamicDirective = deprecatedDynamicDirective !== null && deprecatedDynamicDirective !== void 0 ? deprecatedDynamicDirective : Boolean(experimentalDynamicComponent);
    let result;
    try {
        result = templateCompiler.compile(src, {
            experimentalDynamicDirective,
            // TODO [#3370]: remove experimental template expression flag
            experimentalComplexExpressions,
            preserveHtmlComments,
            enableStaticContentOptimization,
            customRendererConfig,
            enableDynamicComponents,
            instrumentation,
            apiVersion,
        });
    }
    catch (e) {
        throw errors.normalizeToCompilerError(errors.TransformerErrors.HTML_TRANSFORMER_ERROR, e, { filename });
    }
    const fatalError = result.warnings.find((warning) => warning.level === errors.DiagnosticLevel.Error);
    if (fatalError) {
        throw errors.CompilerError.from(fatalError, { filename });
    }
    // The "Error" diagnostic level makes no sense to include here, because it would already have been
    // thrown above. As for "Log" and "Fatal", they are currently unused.
    const warnings = result.warnings.filter((_) => _.level === errors.DiagnosticLevel.Warning);
    const scopeToken = generateScopeToken(filename, namespace, name, apiVersion);
    // Rollup only cares about the mappings property on the map. Since producing a source map for
    // the template doesn't make sense, the transform returns an empty mappings.
    return {
        code: serialize(result.code, filename, scopeToken),
        map: { mappings: '' },
        warnings,
        cssScopeTokens: [
            scopeToken,
            `${scopeToken}-host`, // implicit scope token created by `makeHostToken()` in `@lwc/engine-core`
        ],
    };
}
// The reason this hash code implementation [1] is chosen is because:
// 1. It has a very low hash collision rate - testing a list of 466,551 English words [2], it generates no collisions
// 2. It is fast - it can hash those 466k words in 70ms (Node 16, 2020 MacBook Pro)
// 3. The output size is reasonable (32-bit - this can be base-32 encoded at 10-11 characters)
//
// Also note that the reason we're hashing rather than generating a random number is because
// we want the output to be predictable given the input, which helps with caching.
//
// [1]: https://stackoverflow.com/a/52171480
// [2]: https://github.com/dwyl/english-words/blob/a77cb15f4f5beb59c15b945f2415328a6b33c3b0/words.txt
function generateHashCode(str) {
    const seed = 0;
    let h1 = 0xdeadbeef ^ seed;
    let h2 = 0x41c6ce57 ^ seed;
    for (let i = 0, ch; i < str.length; i++) {
        ch = str.charCodeAt(i);
        h1 = Math.imul(h1 ^ ch, 2654435761);
        h2 = Math.imul(h2 ^ ch, 1597334677);
    }
    h1 = Math.imul(h1 ^ (h1 >>> 16), 2246822507);
    h1 ^= Math.imul(h2 ^ (h2 >>> 13), 3266489909);
    h2 = Math.imul(h2 ^ (h2 >>> 16), 2246822507);
    h2 ^= Math.imul(h1 ^ (h1 >>> 13), 3266489909);
    return 4294967296 * (2097151 & h2) + (h1 >>> 0);
}
function escapeScopeToken(input) {
    // Minimal escape for strings containing the "@" and "#" characters, which are disallowed
    // in certain cases in attribute names
    return input.replace(/@/g, '___at___').replace(/#/g, '___hash___');
}
function generateScopeToken(filename, namespace, name, apiVersion) {
    const uniqueToken = `${namespace}-${name}_${path__namespace.basename(filename, path__namespace.extname(filename))}`;
    if (shared.isAPIFeatureEnabled(0 /* APIFeature.LOWERCASE_SCOPE_TOKENS */, apiVersion)) {
        const hashCode = generateHashCode(uniqueToken);
        // This scope token is all lowercase so that it works correctly in case-sensitive namespaces (e.g. SVG).
        // It is deliberately designed to discourage people from relying on it by appearing somewhat random.
        // (But not totally random, because it's nice to have stable scope tokens for our own tests.)
        // Base-32 is chosen because it is not case-sensitive (0-v), and generates short strings with the given hash
        // code implementation (10-11 characters).
        return `lwc-${hashCode.toString(32)}`;
    }
    else {
        // This scope token is based on the namespace and name, and contains a mix of uppercase/lowercase chars
        return escapeScopeToken(uniqueToken);
    }
}
function serialize(code, filename, scopeToken) {
    const cssRelPath = `./${path__namespace.basename(filename, path__namespace.extname(filename))}.css`;
    const scopedCssRelPath = `./${path__namespace.basename(filename, path__namespace.extname(filename))}.scoped.css`;
    let buffer = '';
    buffer += `import { freezeTemplate } from "lwc";\n\n`;
    buffer += `import _implicitStylesheets from "${cssRelPath}";\n\n`;
    buffer += `import _implicitScopedStylesheets from "${scopedCssRelPath}?scoped=true";\n\n`;
    buffer += code;
    buffer += '\n\n';
    buffer += 'if (_implicitStylesheets) {\n';
    buffer += `  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitStylesheets);\n`;
    buffer += `}\n`;
    buffer += 'if (_implicitScopedStylesheets) {\n';
    buffer += `  tmpl.stylesheets.push.apply(tmpl.stylesheets, _implicitScopedStylesheets);\n`;
    buffer += `}\n`;
    buffer += `tmpl.stylesheetToken = "${scopeToken}";\n`;
    // Note that `renderMode` and `slots` are already rendered in @lwc/template-compiler and appear
    // as `code` above. At this point, no more expando props should be added to `tmpl`.
    buffer += 'freezeTemplate(tmpl);\n';
    return buffer;
}

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
function scriptTransform(code, filename, options) {
    const { isExplicitImport, experimentalDynamicComponent: dynamicImports, outputConfig: { sourcemap }, enableLightningWebSecurityTransforms, namespace, name, instrumentation, apiVersion, } = options;
    const lwcBabelPluginOptions = {
        isExplicitImport,
        dynamicImports,
        namespace,
        name,
        instrumentation,
        apiVersion,
    };
    const plugins = [
        [lwcClassTransformPlugin, lwcBabelPluginOptions],
        [babelClassPropertiesPlugin, { loose: true }],
        // This plugin should be removed in a future version. The object-rest-spread is
        // already a stage 4 feature. The LWC compile should leave this syntax untouched.
        babelObjectRestSpreadPlugin,
    ];
    if (enableLightningWebSecurityTransforms) {
        plugins.push(lockerBabelPluginTransformUnforgeables, babelAsyncToGenPlugin, babelAsyncGeneratorFunctionsPlugin);
    }
    let result;
    try {
        result = babel__namespace.transformSync(code, {
            filename,
            sourceMaps: sourcemap,
            // Prevent Babel from loading local configuration.
            babelrc: false,
            configFile: false,
            // Force Babel to generate new line and white spaces. This prevent Babel from generating
            // an error when the generated code is over 500KB.
            compact: false,
            plugins,
        });
    }
    catch (e) {
        throw errors.normalizeToCompilerError(errors.TransformerErrors.JS_TRANSFORMER_ERROR, e, { filename });
    }
    return {
        code: result.code,
        map: result.map,
    };
}

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
/**
 * Transforms the passed code. Returning a Promise of an object with the generated code, source map
 * and gathered metadata.
 *
 * @deprecated Use transformSync instead.
 */
function transform(src, filename, options) {
    validateArguments(src, filename);
    return new Promise((resolve, reject) => {
        try {
            const res = transformSync(src, filename, options);
            resolve(res);
        }
        catch (error) {
            reject(error);
        }
    });
}
/**
 * Transform the passed source code. Returning an object with the generated code, source map and
 * gathered metadata.
 */
function transformSync(src, filename, options) {
    validateArguments(src, filename);
    const normalizedOptions = validateTransformOptions(options);
    return transformFile(src, filename, normalizedOptions);
}
function validateArguments(src, filename) {
    errors.invariant(shared.isString(src), errors.TransformerErrors.INVALID_SOURCE, [src]);
    errors.invariant(shared.isString(filename), errors.TransformerErrors.INVALID_ID, [filename]);
}
function transformFile(src, filename, options) {
    let transformer;
    switch (path__namespace.extname(filename)) {
        case '.html':
            transformer = templateTransform;
            break;
        case '.css':
            transformer = styleTransform;
            break;
        case '.ts':
        case '.js':
            transformer = scriptTransform;
            break;
        default:
            throw errors.generateCompilerError(errors.TransformerErrors.NO_AVAILABLE_TRANSFORMER, {
                messageArgs: [filename],
                origin: { filename },
            });
    }
    return transformer(src, filename, options);
}

/*
 * Copyright (c) 2018, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */
const version = "3.0.0";

exports.transform = transform;
exports.transformSync = transformSync;
exports.version = version;
/** version: 3.0.0 */
